### DISCUSSION TITLE
Feature ideas and work

### DISCUSSION THREAD

**Discussion by benpayne** [EUdds/Digikey-Inventree-Integration#12] [Ideas]
This thread is intended to discuss possible feature enhancements that I've been looking at and working on.  So far you've seen a few of the things that I've added.

1. Automatic Category assignment
2. Importing Parameters from DK
3. Importing of HTSUS codes as a Parameter in Inventory
4. Importing Orders from DK as POs in Inventree

Other ideas that I'm looking at adding:
1. Importing BOM from octopart to part in Inventree
2. Exporting Purchase order from Inventree to DK cart
3. Add support for working with Mouser (so do everything that we can do with DK)
4. Adding links to octopart for all parts.  
5. Exporting BOM to octopart, will help find current suppliers for parts if main distributor is out.
6. Add web UI to tool and/or integrate into Invetree as a plugin.  
7. Probably a few more I'm forgetting right now...

For simplicity I've used the config file as the source to control the operations, like Category/Parameters mapping.  Are you onboard with using the configuration file in this way?  I had thought that we could map these in a python file, or we could have them in a separate config file.  The ini format was a bit counter intuitive at first to describe this data, but in the end I think it's fairly good.  

I've continued to expand on the command line interface that you created, but also looked to add support that these actions could be done by using this project as a python module and controlling this from your own external code.  Do we want to support both use cases?  

What other ideas have you had?  It would be good to understand what your vision is.  Don't want to design anything that would prevent your ideas or be in contrast.   

Look forward to discussing this more, this project is a great start and certainly jumpstarted my work, it's always great to have a working base to build from.  

Thanks

**Comment by EUdds** [comment]
Hey,

Let me give the context behind this project so I'm upfront about what I'm thinking.

I built this project to help my undergrad solar car team import digikey parts into Inventree to reduce busy work. I never intended for this to be used by more than like 3 people who had me within earshot so I could fix things if they broke. So, the code is not great. **That being said** I think it's super cool that people are using it, I want to clean this up and make it even more useful so more people can enjoy it. So as far as my plan goes, I have no plan, I am ready to come up with a roadmap in this thread.

So let me tackle the OP in a few parts here in order of which I think they could get done

2. This seems like the easiest, least amount of work for the most amount of feature benefit

3. We can definitely do this, if we are going to do this, we should probably refactor the whole DKPart class and have a base class that each supplier can be extended off of. If we're going to do this, let's do it right because...

1,4, and 5. are things that I have been wanting to do but never had the drive to do. I think that this is entirely doable but I would really want to do the "Add other suppliers" feature first so that we come up with a backbone to put this on.

6. We _can_ do, but Inventree now has native plugin support, it might just be best to develop and offshoot project that utilizes this one that's an Inventree plugin

Here are some ideas that I would like to add (also in order of which I think they should get done)
1. Global config module, allows the user to specify the path of a config file, and/or populate the config programmatically (Being worked on in #11 )
2. Package the project. I'm a fan of poetry, but could be convinced to revert to standard `setuptools` , no super strong preference just yet. (and also publish on PIP)
3. End-to-End testing, I want this to be reliable, the unit tests I spun up are good at boding confidence, but I think some end to end work with an Inventree server in the runner would be sick

Let me know what you think! I'm happy to workshop a bit, I'm glad that someone has found use in this project and is willing to help contribute :)

Cheers!

**Comment by EUdds** [comment]
Hey @benpayne , I revamped the package into a poetry package, would love to get your eyes on this PR: https://github.com/EUdds/digikey-inventree-integration/pull/13 thanks!

**Comment by benpayne** [comment]
Thanks for all the comments.  I think we have similar reasons for our interest in the project, we are using it to track inverts at our small mostly prototype production house, mostly for in-house designs.  Inventree was recently deployed to replace out previous tool and these tools have been critical to helping get data imported.  I've been busy the last week or so with other work, but should have some time next week to look more at the work you did.  

I will say that I've been thinking that we should wrap the code in Inventree.py in a class, would simplify the initialization int the constructor of the class.  I've also been thinking about how to build the code around introducing another supplier (aka Mouser).  Like you said we need some base classes.  I think it also means that some of the code in Inventree.py maybe should move over to the [supplier].py file.  Like finding and creating its supplier record in inventree.  With this in mind there shouldn't be any code in Inventree.py that is specific to one supplier.  Previously I'd been thinking that it was clean to not import Inventree api into Digikey.py, but now I'm thinking that it does.  If you think about the project overall it's all about getting data in/out of Inventee so why not.  But I did also think that there could be low-level classes that just wrap the supplier specific code and then those class could be used in a new layer that pulls in the Inventree api.  Something like this:

Inventree.py -uses->  SupplierBase.py -is a-> SupplierDigikey.py -uses-> Digikey.py

But my current thought is that SupplierDigikey.py and Digikey.py could just be the same file/classes.  Let me know if you have thoughts on this.

**Reply by EUdds** [reply]
I think some minor semantics things about Inventree.py but we can cross that when we get there, overall I was thinking the same there

**Reply by EUdds** [reply]
Hey @benpayne  - stubbed some of this out here: https://github.com/EUdds/digikey-inventree-integration/pull/23 lmk what you think!

**Comment by EUdds** [comment]
Hey so I recently setup the full package structure, so I think we're good to start object orienting the supplier details. LMK what you think about the recent poetry changes.
